### PR TITLE
Strict typing

### PR DESCRIPTION
**Pull request by aslakhellesoy**
Thanks for this great little library!

This PR changes the TypeScript typings to always assume that a field may be `null` or `undefined`. It also declares all types and fields as immutable.

The rationale behind this change is that in practice, we cannot safely assume anything about the original XML document. This is because it isn't validated against an XML Schema or DTD before it's translated into JSON. The lack of an official schema for JUnit XML means that we cannot take anything for granted  - different tools may produce XML that lacks properties.

The proposed type changes helps users of this library anticipate that a field may be `null` or `undefined` (assuming they have set `strictNullChecks` to `true` in `tsconfig.json`. I also did that for this library, and fixed some of the tests.

### CODE REVIEW

**Comment by Kesin11** [comment]
That's an interesting suggestion.
I have only been applying security patches for a long time, give me some time to make my opinion.

**Comment by Kesin11** [comment]
I respected `TestSuites`, `TestCase` and `TestSuite` types following these document.
https://github.com/junit-team/junit5/blob/main/platform-tests/src/test/resources/jenkins-junit.xsd
https://llg.cubic.org/docs/junit/

But as you commented, XML generated from any of the test tools may not follow these schemas. So I agree to treat all of field as optional types. However, I disagree about `Maybe`(1) and `Readonly`(2) in your changes.

(1) Which situation does `_parse()` produce `null` ? Even if the input XML tag is empty, the output value should be undefined or empty string.

(2) So 'parse()' returns a plain object, `Readonly` is not correct. If users want to treat obect as readonly, use some type utilities such as `ReadonlyDeep` from convenient type package (ex [sindresorhus/type-fest](https://github.com/sindresorhus/type-fest)) and should cast type in yourself.

File: __tests__/basic.test.ts [hunk 1]
```diff
@@ -29,15 +29,17 @@ describe('Convert xml2js output tests', () => {
...
-    })
+    } as junit2json.TestSuites)
   })
 
   it('array with string array', async () => {
     const xml = `
     <?xml version="1.0" encoding="UTF-8"?>
     <testsuites>
       <testsuite tests="1">
-        <failure>inner text</failure>
```
Comment (Kesin11) [inline __tests__/basic.test.ts:40]: Oh, these are my mistake. thanks to fix 👍

File: src/index.ts [hunk 2]
```diff
@@ -1,56 +1,62 @@
...
+export type Details = Readonly<{ message?: Maybe<string>, type?: Maybe<string>, inner?: Maybe<string> }>
+
+export const parse = async (xmlString: xml2js.convertableToString, xml2jsOptions?: xml2js.OptionsV2): Promise<TestSuites|TestSuite|undefined> => {
   const options = xml2jsOptions ?? {
     attrValueProcessors: [xml2js.processors.parseNumbers]
   }
   const result = await xml2js.parseStringPromise(xmlString, options)
 
-  if (result['testsuites']) {
+  if ('testsuites' in result) {
```
Comment (Kesin11) [inline src/index.ts:56]: 👍

File: src/index.ts [hunk 3]
```diff
@@ -1,56 +1,62 @@
...
+  properties?: Maybe<readonly Property[]>
+  "system-out"?: Maybe<readonly string[]>
+  "system-err"?: Maybe<readonly string[]>
+}>
+
+export type Property = Readonly<{ name?: Maybe<string>, value?: Maybe<string> }>
+export type Skipped = Readonly<{ message?: Maybe<string> }>
+export type Details = Readonly<{ message?: Maybe<string>, type?: Maybe<string>, inner?: Maybe<string> }>
+
+export const parse = async (xmlString: xml2js.convertableToString, xml2jsOptions?: xml2js.OptionsV2): Promise<TestSuites|TestSuite|undefined> => {
```
Comment (Kesin11) [inline src/index.ts:50]: Thanks to add `undefined`. I'll add some tests when input XML is invalid.

**Comment by Kesin11** [comment]
hmm... Why CI did not trigger this pull request? 🤔 
I'll check Github Actions yaml and fixit.

**Comment by aslakhellesoy** [comment]
> I respected `TestSuites`, `TestCase` and `TestSuite` types following these document. https://github.com/junit-team/junit5/blob/main/platform-tests/src/test/resources/jenkins-junit.xsd https://llg.cubic.org/docs/junit/
> 
> But as you commented, XML generated from any of the test tools may not follow these schemas. So I agree to treat all of field as optional types. 

👍🏻 

> However, I disagree about `Maybe`(1) and `Readonly`(2) in your changes.
> 

Good point. The use of `readonly` is not needed. I will update the PR and remove that.

> (1) Which situation does `_parse()` produce `null` ? Even if the input XML tag is empty, the output value should be undefined or empty string.
> 

I'll fix this too.

> (2) So 'parse()' returns a plain object, `Readonly` is not correct. If users want to treat obect as readonly, use some type utilities such as `ReadonlyDeep` from convenient type package (ex [sindresorhus/type-fest](https://github.com/sindresorhus/type-fest)) and should cast type in yourself.

👍🏻 will update

**Comment by aslakhellesoy** [comment]
> hmm... Why CI did not trigger this pull request? 🤔 I'll check Github Actions yaml and fixit.

I think you need to approve it manually:
Screenshot 2022-01-31 at 10 35 06

File: src/index.ts (hunk 3 above)
Comment (aslakhellesoy) [reply src/index.ts]: I've added two more tests that verify what happens with an empty XML document as well as an XML document with an unrecognized root element.

**Comment by aslakhellesoy** [comment]
Thanks for the feedback and sorry for the late reply! I hope the recent changes addresses all your concerns :-)

File: src/index.ts [hunk 4]
```diff
@@ -1,56 +1,63 @@
 import xml2js from 'xml2js'
 
+type Maybe<T> = T | null
```
Comment (Kesin11) [inline src/index.ts:3]: I just found `parse()` can return null. However, does `TestSuites` and the `TestSuite` property can be null?
I think it is correct to make them optional types, but adding `Maybe` would make them different types than they really are.

If the XML does not have a certain key, then the returned object does not have the same key.
And even if the XML has a key with a value of null, the returned object will have a value of String "null". 

```js
  it('tests="null" is impossible case', async () => {
    const xml = `
    <?xml version="1.0" encoding="UTF-8"?>
    <testsuites>
      <testsuite failures="1" tests="null">
      </testsuite>
    </testsuites>
    `
    const parsed = await junit2json.parse(xml)

    expect(parsed).toEqual({
      testsuite: [
        { failures: 1, tests: "null" }, // It convert to String "null"
      ]
    })
  })
```

Can you add a test if you know some case returned values include null?

**Review by Kesin11 — COMMENTED** [review]
Thanks to fix codes and add tests.
However, could you please show me when properties are null?

**Comment by aslakhellesoy** [comment]
> @aslakhellesoy Thanks to fix codes and add tests. However, could you please show me when properties are null?

My mistake - they cannot be `null`. If they are absent in the XML they are `undefined`. I've removed the `Maybe` type and added a test to verify this.

**Review by Kesin11 — APPROVED** [review]
LGTM 👍 
Thanks!

**Comment by Kesin11** [comment]
@aslakhellesoy 
I've released v3.0.0 🎉 
https://github.com/Kesin11/ts-junit2json/releases/tag/v3.0.0

**Comment by aslakhellesoy** [comment]
Thanks a lot @Kesin11 - and thanks for the great feedback on the PR too.
